shared-module/wiznet/wiznet5k.c

```diff
@@ -0,0 +1,403 @@
+    // TODO convert Wiz errno's to POSIX ones
+    if (ret < 0) {
+        wiznet5k_socket_close(socket);
+        *_errno = -ret;
+        return -1;
+    }
+    return ret;
+}
+
+mp_uint_t wiznet5k_socket_sendto(mod_network_socket_obj_t *socket, const byte *buf, mp_uint_t len, byte *ip, mp_uint_t port, int *_errno) {
+    if (socket->u_param.domain == 0) {
+        // socket not opened; use "bind" function to open the socket in client mode
+        if (wiznet5k_socket_bind(socket, ip, 0, _errno) != 0) {
+            return -1;
+        }
+    }
+
+    MP_THREAD_GIL_EXIT();
+    mp_int_t ret = WIZCHIP_EXPORT(sendto)(socket->u_param.fileno, (byte*)buf, len, ip, port);
+    MP_THREAD_GIL_ENTER();
+
+    if (ret < 0) {
+        wiznet5k_socket_close(socket);
+        *_errno = -ret;
+        return -1;
+    }
+    return ret;
+}
+
+mp_uint_t wiznet5k_socket_recvfrom(mod_network_socket_obj_t *socket, byte *buf, mp_uint_t len, byte *ip, mp_uint_t *port, int *_errno) {
+    uint16_t port2;
+    MP_THREAD_GIL_EXIT();
+    mp_int_t ret = WIZCHIP_EXPORT(recvfrom)(socket->u_param.fileno, buf, len, ip, &port2);
+    MP_THREAD_GIL_ENTER();
+    *port = port2;
+    if (ret < 0) {
+        wiznet5k_socket_close(socket);
+        *_errno = -ret;
+        return -1;
+    }
+    return ret;
+}
+
+int wiznet5k_socket_setsockopt(mod_network_socket_obj_t *socket, mp_uint_t level, mp_uint_t opt, const void *optval, mp_uint_t optlen, int *_errno) {
+    // TODO
+    *_errno = MP_EINVAL;
+    return -1;
+}
+
+int wiznet5k_socket_settimeout(mod_network_socket_obj_t *socket, mp_uint_t timeout_ms, int *_errno) {
+    // TODO
+    *_errno = MP_EINVAL;
+    return -1;
+
+    /*
+    if (timeout_ms == 0) {
+        // set non-blocking mode
+        uint8_t arg = SOCK_IO_NONBLOCK;
+        WIZCHIP_EXPORT(ctlsocket)(socket->u_param.fileno, CS_SET_IOMODE, &arg);
+    }
+    */
+}
+
+int wiznet5k_socket_ioctl(mod_network_socket_obj_t *socket, mp_uint_t request, mp_uint_t arg, int *_errno) {
+    if (request == MP_STREAM_POLL) {
+        int ret = 0;
+        if (arg & MP_STREAM_POLL_RD && getSn_RX_RSR(socket->u_param.fileno) != 0) {
+            ret |= MP_STREAM_POLL_RD;
+        }
+        if (arg & MP_STREAM_POLL_WR && getSn_TX_FSR(socket->u_param.fileno) != 0) {
+            ret |= MP_STREAM_POLL_WR;
+        }
+        return ret;
+    } else {
+        *_errno = MP_EINVAL;
+        return MP_STREAM_ERROR;
+    }
+}
+
+static void wiznet5k_try_dhcp(void) {
+    DHCP_INIT_BUFFER_TYPE dhcp_buf[DHCP_INIT_BUFFER_SIZE];
+
+    // Set up the socket to listen on UDP 68 before calling DHCP_init
+    WIZCHIP_EXPORT(socket)(0, MOD_NETWORK_SOCK_DGRAM, DHCP_CLIENT_PORT, 0);
+    DHCP_init(0, dhcp_buf);
+
+    // try a few times for DHCP ... XXX this should be asynchronous.
+    for (int i=0; i<10; i++) {
+        DHCP_time_handler();
+        int dhcp_state = DHCP_run();
+        if (dhcp_state == DHCP_IP_LEASED || dhcp_state == DHCP_IP_CHANGED) break;
+        mp_hal_delay_ms(1000);
```

Comment:
This makes the construction of the object really slow when not plugged in. Can we do one try in the constructor and one each time the other APIs are called?